apps/tidesandtemp/tides_and_temp.star

```diff
@@ -0,0 +1,202 @@
+"""
+Applet: Tides & Temp
+Summary: Display tides / water temp
+Description: Display predicted tides from NOAA tide stations and water temperature from NDBC buoy stations. See https://tidesandcurrents.noaa.gov/tide_predictions.html for NOAA tide station IDs. See https://www.ndbc.noaa.gov/obs.shtml for NDBC buoy station IDs.
+Author: sudeepban
+"""
+
+load("encoding/base64.star", "base64")
+load("http.star", "http")
+load("render.star", "render")
+load("schema.star", "schema")
+load("time.star", "time")
+
+NOAA_TIDES_URL = "https://api.tidesandcurrents.noaa.gov/api/prod/datagetter?station={station_id}&product=predictions&datum=MLLW&time_zone=gmt&interval=hilo&units=english&application=DataAPI_Sample&format=json&range=48&begin_date={begin_date}"
+NOAA_TIDES_STATION_ID_DEFAULT = 8533071
+
+NDBC_BUOY_URL = "https://www.ndbc.noaa.gov/data/realtime2/{station_id}.txt"
+NDBC_BUOY_STATION_ID_DEFAULT = 44091
+
+LOW_TIDE_COLOR_DEFAULT = "#088F8F"
+HIGH_TIDE_COLOR_DEFAULT = "#1F51FF"
+WATER_TEMPERATURE_COLOR_DEFAULT = "#EEDC82"
+
+TIMEZONE_DEFAULT = "America/New_York"
+TIMEZONE_GMT = "GMT"
+
+DATE_FORMAT = "20060102"
+DATETIME_FORMAT = "2006-01-02 15:04"
+TIME_FORMAT = "03:04 PM"
+
+def main(config):
+    timezone = config.get("timezone", TIMEZONE_DEFAULT)
+    noaaTidesStationID = config.str("noaaTidesStationID", str(NOAA_TIDES_STATION_ID_DEFAULT))
+    ndbcBuoyStationID = config.str("ndbcBuoyStationID", str(NDBC_BUOY_STATION_ID_DEFAULT))
+    lowTideColor = config.str("lowTideColor", LOW_TIDE_COLOR_DEFAULT)
+    highTideColor = config.str("highTideColor", HIGH_TIDE_COLOR_DEFAULT)
+    waterTempColor = config.str("waterTempColor", WATER_TEMPERATURE_COLOR_DEFAULT)
+
+    now = time.now().in_location(TIMEZONE_GMT)
+    prevdate = (now - time.parse_duration("12h")).format(DATE_FORMAT)
+
+    resp = http.get(NOAA_TIDES_URL.format(begin_date = prevdate, station_id = noaaTidesStationID))
```

Comment:
Would it be appropriate to add a `ttl_seconds` flag to this call? Not sure if they provide cache-control headers?

Reply:
I didn't see anything at TTL or cache-control headers at https://api.tidesandcurrents.noaa.gov/api/prod/ - let me know if you see anything relevant I could incorporate

Reply:
The simplest thing is probably for you to decide how long the data retrieved is good for. So for instance, if it's ok to use the buoy data for (say) 1 hour, you could do `http.get(NDBC_BUOY_URL.format(station_id = ndbcBuoyStationID), ttl_seconds = 3600)`. This means subsequent requests for the exact same URL will not hit the NOAA server, but will instead be served from our cache. After 3600 seconds (=1h) the cached data will get thrown out and the next request will refresh the data from the API again.

Reply:
I see, that makes sense. My preference would be for it to be up to date, with the latest water temperature reading as soon as it's made available. It seems like adding any non-zero `ttl_seconds` would introduce some potential arbitrary delay.

Reply:
Totally understand wanting up to date data. That makes sense. Unfortunately we do have to cache responses to keep request volumes at a resonable level.

I took a look at the data served on the buoy url, and it includes cache-control headers requesting 600s caching of responses, so that's the minimum we should do there. In fact, that's what'll happen automatically when `ttl_seconds` is left out.

There doesn't seem to be such a header on the other endpoint, but if I understand this correctly, it only serves tide predictions. These likely do not change much in the short term, so an hour of caching shouldn't hurt I imagine?

For reference, the existing apps nooatides and nooabuoy cache responses for 4 hours and 10 minute respectively.

Reply:
Thanks for the details.  I accepted your proposed changes, this should be fine!